### ISSUE TITLE
Loss output gives Nan in Linux but gives normal values in Windows 

### ISSUE THREAD

**Issue by sliu729** [tensorflow/tensorflow#52068]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux & Windows 10
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: No
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 2.4.1
- Python version: 3.8.10
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source): GCC 7.5.0
- CUDA/cuDNN version:10.1
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
I defined a model (two inputs and two outputs) that trains well on my local PC (Windows 10), but only produce Nan in Linux
**Describe the expected behavior**
The model should be trained properly in both Windows and Linux

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no): no
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.
[slurm-8183898.txt](https://github.com/tensorflow/tensorflow/files/7195723/slurm-8183898.txt)

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by vulkomilev** [comment]
can you provide the code ?

**Comment by sushreebarsa** [comment]
@sliu729 
In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thanks!

**Comment by sliu729** [comment]
@vulkomilev  @sushreebarsa Hey, thanks for your prompt reaction. The code was not organized to be portable, I will try to reorganize and provide the code to reproduce the problem. In the meanwhile, the NaN problem disappears when I disabled the use of mixed_float16. Is that something already reported by someone else?

**Comment by vulkomilev** [comment]
Maybe the problem was the usage of different types of variables for weights and backprop. The float format is not very intuitive and funny things can happen.

**Comment by sliu729** [comment]
> Maybe the problem was the usage of different types of variables for weights and backprop. The float format is not very intuitive and funny things can happen.

Thanks for your comment. The funny thing is that I got recommended by the profiler to use mixed_float16, and the calculation speed was greatly increased. It is sad that it would not work properly.

**Comment by sushreebarsa** [comment]
@sliu729 
In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thanks!
